freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UsersyncPoller.java

```diff
@@ -113,13 +99,67 @@ void syncFreeIpaStacks() {
+                } catch (IOException e) {
+                    LOGGER.warn("Failed to retrieve UmsEventGenerationIds for Environment {} in Account {}. Assuming stale",
+                            stack.getEnvironmentCrn(), stack.getAccountId());
+                    stale = true;
+                }
+            }
+            LOGGER.debug("Environment {} in Account {} {} stale", stack.getEnvironmentCrn(), stack.getAccountId(), stale ? "is" : "is not");
+            return stale;
         } catch (Exception e) {
             LOGGER.warn("Unable to calculate staleness due to exception.", e);
             return false;
         }
     }
+
+    @VisibleForTesting
+    boolean isCooldownExpired(Stack stack, UserSyncStatus userSyncStatus, long cooldownThresholdTime) {
+        try {
+            boolean cool = userSyncStatus == null ||
+                    userSyncStatus.getLastFullSyncStartTime() == null ||
+                    userSyncStatus.getLastFullSyncStartTime() < cooldownThresholdTime;
+            LOGGER.debug("Environment {} in Account {} {} cool", stack.getEnvironmentCrn(), stack.getAccountId(), cool ? "is" : "is not");
+            return cool;
+        } catch (Exception e) {
+            LOGGER.warn("Unable to determine if cooldown expired due to exception.", e);
+            return false;
+        }
+    }
+
+    private boolean isAccountEntitled(String accountId) {
+        boolean entitled = entitlementService.automaticUsersyncPollerEnabled(INTERNAL_ACTOR_CRN, accountId);
+        if (!entitled) {
+            LOGGER.debug("Usersync polling is not entitled in accout {}. skipping", accountId);
+        }
+        return entitled;
+    }
+
+    private void syncFreeIpaStacksInAccount(Optional<String> requestId, String accountId, List<Stack> stacks) {
+        long cooldownThresholdTime = System.currentTimeMillis() - cooldownMillis;
+
+        UmsEventGenerationIds currentGeneration =
+                umsEventGenerationIdsProvider.getEventGenerationIds(accountId, requestId);
+        stacks.stream()
+                .forEach(stack -> {
+                    UserSyncStatus userSyncStatus = userSyncStatusService.getOrCreateForStack(stack);
+                    if (isStale(stack, userSyncStatus, currentGeneration) && isCooldownExpired(stack, userSyncStatus, cooldownThresholdTime)) {
+                        LOGGER.debug("Environment {} in Account {} is both stale and cool.",
+                                stack.getEnvironmentCrn(), stack.getAccountId());
+                        SyncOperationStatus operation = userSyncService.synchronizeUsers(stack.getAccountId(), INTERNAL_ACTOR_CRN,
+                                Set.of(stack.getEnvironmentCrn()), Set.of(), Set.of());
+                        LOGGER.debug("User Sync request resulted in operation {}", operation);
+                    } else {
+                        LOGGER.debug("Environment {} in Account {} is up-to-date.", stack.getEnvironmentCrn(), stack.getAccountId());
```

Comment:
this is not 100% precise, it can be "hot".

Reply:
not cool, dude. not cool....